WcaOnRails/lib/results_validators/advancement_conditions_validator.rb

```diff
@@ -85,16 +112,43 @@ def validate(competition_ids: [], model: Result, results: nil)
                                                  round_id: round_id)
                 end
               else
+                max_advancing = 3 * number_of_people_in_previous_round / 4
                 # Article 9p1, since April 14, 2010
                 # https://www.worldcubeassociation.org/regulations/#9p1: At least 25% of competitors must be eliminated between consecutive rounds of the same event.
-                if number_of_people_in_round > 3 * number_of_people_in_previous_round / 4
+                if number_of_people_in_round > max_advancing
                   @errors << ValidationError.new(:rounds, competition_id,
                                                  REGULATION_9P1_ERROR,
                                                  round_id: round_id)
                 end
+                if condition
+                  theoritical_number_of_people = condition.max_advancing(previous_results)
+                  if number_of_people_in_round > theoritical_number_of_people
+                    @warnings << ValidationWarning.new(:rounds, competition_id,
+                                                       TOO_MANY_QUALIFIED_WARNING,
+                                                       round_id: round_id,
+                                                       actual: number_of_people_in_round,
+                                                       expected: theoritical_number_of_people,
+                                                       condition: condition.to_s(previous_round))
+                  end
+                  if theoritical_number_of_people > max_advancing
+                    @errors << ValidationError.new(:rounds, competition_id,
+                                                   ROUND_9P1_ERROR,
+                                                   round_id: round_id,
+                                                   condition: condition.to_s(previous_round))
+                  end
+                  # This comes from https://github.com/thewca/worldcubeassociation.org/issues/5587
+                  if theoritical_number_of_people - number_of_people_in_round >= 3 &&
+                     (number_of_people_in_round / theoritical_number_of_people) <= 0.8
```

Comment:
I noticed this warning while submitting CubingUSANationals2023:

> Round 333-2: according to the events data, at most 400 could have proceed, but only 395 competed in the round. Please leave a comment about that (or fix the events data if you applied a different advancement condition).
> Round 555-2: according to the events data, at most 100 could have proceed, but only 96 competed in the round. Please leave a comment about that (or fix the events data if you applied a different advancement condition).
> Round pyram-2: according to the events data, at most 100 could have proceed, but only 96 competed in the round. Please leave a comment about that (or fix the events data if you applied a different advancement condition).
> Round skewb-2: according to the events data, at most 100 could have proceed, but only 96 competed in the round. Please leave a comment about that (or fix the events data if you applied a different advancement condition).

I think this isn't actually doing the second check because of integer division -- the left side of the equation is always 0 if number_of_people_in_round < theoretical_number_of_people.